Merge odometry logic

vecless opened the pull request:
Branch name is a bit of a misnomer, this doesn't introduce any new threads or Notifiers anymore (for simplicity reasons).

Probably needs code review because it's been a while since I looked at this last, tbh.

a-viv-a committed:
Merge branch 'dev' into main

vecless committed:
Add files

vecless committed:
Add SimpleSwerveTrajectoryFollower

vecless committed:
Add FollowTrajectoryCommand

vecless committed:
Add example use of FollowTrajectoryCommand into RobotContainer

a-viv-a committed:
dry the drivebase for tab creation

vecless committed:
Split periodic steps remove unnecessary null check

vecless committed:
Merge remote-tracking branch 'origin/feat/swerve-update-thread' into feat/swerve-update-thread

vecless committed:
Label members of DrivebaseSubsystem as Guarded

vecless committed:
Final commit before de-thread

vecless committed:
Upload config from video on 2022-01-27

Check slack channel for relevant video. Also removes miscellaneous warnings which have been adequately addressed, as far as I know

vecless committed:
Re-remove faster update loopfor better reading experience

vecless committed:
Remove SDS Swerve Lib from source, add as vendordep

Woohoo! Finally moving to the new 2022 release of the SDS swerve-lib

vecless committed:
Merge remote-tracking branch 'origin/dev' into feat/swerve-update-thread

vecless committed:
Handle intersecting names

todo: still need to fix pathing of lib stuff (I'll just move it all to frc.util pkg)

a-viv-a commented on build.gradle:
```diff
@@ -4,6 +4,7 @@ plugins {
     id "java"
     id "edu.wpi.first.GradleRIO" version "2022.2.1"
     id "com.diffplug.spotless" version "6.2.0"
+    id "net.ltgt.errorprone" version "2.0.2"
```
do we still need the errorprone compiler?

vecless replied:
Oops, you're right, we can drop this provided all the GuardedBy annotations are dropped

a-viv-a commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -17,17 +17,15 @@
   private Command autonomousCommand;
 
-  private RobotContainer robotContainer;
+  private RobotContainer robotContainer = new RobotContainer();
```
Is there a reason to do this before ``robotInit``? Supposedly the ``robotInit`` is the right function for initialization code?

vecless replied:
I don't remember why I did this. I think you're right, let's change it back to that

a-viv-a commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -61,6 +70,7 @@ private void configureButtonBindings() {
 
     new Button(nick::getAButton).whenPressed(drivebaseSubsystem::zeroGyroscope);
     new Button(nick::getLeftBumper).whenHeld(new DefenseModeCommand(drivebaseSubsystem));
+    new Button(will::getAButton).whenPressed(new PrintCommand("will a button"));
```
Is there a reason for this? I assume a debugging line that slipped in - we use the ``a button`` on the rotation branch.

vecless replied:
We can just remove this yeah

a-viv-a commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -88,4 +113,8 @@ private static double modifyAxis(double value) {
     return value;
   }
+
+  protected DrivebaseSubsystem getDrivebaseSubsystem() {
```
Looks like this method is unused - what is it for?

vecless replied:
Ur right. Let's remove

vecless replied:
No use for this given the current structure anymore -- this was in anticipation of something that never ended up happening

a-viv-a commented on src/main/java/frc/robot/autonomous/SimpleSwerveTrajectoryFollower.java:
```diff
@@ -0,0 +1,78 @@
+    double yFF = linearVelocityRefMeters * poseRef.getRotation().getSin();
+    // This sets the fixed angle goal to be 0. This would ideally be changed in a more advanced
+    // follower.
+    double angleFF = angleController.calculate(currentPose.getRotation().getRadians(), 0);
```
This might make sense to integrate with the drive angle stuff, but that's not on dev - thoughts?

vecless replied:
Main use I see in trajectory is during a segment we may want to rotate 90 degrees over course of a path in which case we may want to provide an interpolable heading value for the duration of the path

otherwise I don't see why we wouldn't just used a fixed angle as changing the facing angle during the trajectory seems "sorta useless"

a-viv-a replied:
I meant in the sense we shouldn't duplicate the pid controllers. What do profiled pid controllers offer?

vecless replied:
Ah, I see what you mean.
ProfiledPIDController allows us to constrain setpoints based on a trapezoidal motion profile for how we'd like our angle to change.
One could say that it's nice like this because it might help with autonomous trajectory consistency whereas in teleop intuitive mapping of controls->action is more highly prioritized.
But I do see your idea that these could be good to unify. In the version of `dev` this is built from, there was no existing angle PID controller yet, so that was the main reason it's a separate one.

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+  private final Object trajectoryLock = new Object();
+
+  /** The trajectory that is currently being followed. null if no trajectory is being followed. */
+  @GuardedBy("trajectoryLock")
```
if there are no threads, do we need guards?

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+
+/** Follower for trajectory of type T */
+public abstract class TrajectoryFollower<T> {
+  private final Object trajectoryLock = new Object();
```
If there are no threads, do we need locks?

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+  protected abstract void reset();
+
+  public final void cancel() {
+    synchronized (trajectoryLock) {
```
Is the synchronized block needed?

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+    synchronized (trajectoryLock) {
+      // Return empty if no trajectory is being followed
+      if (currentTrajectory == null) {
+        return Optional.empty();
```
Was optional intended to be used this way? I personally am a fan of this, but it is probably important to standardize. Looks good in this case, just curious.

vecless replied:
You are supposed to use the static Optional.empty() to return an empty instance, yeah. As for convention of use this is pretty standard across langs and also documented so I think it's fine. Open to what you wanna do with this though

a-viv-a commented on src/main/java/frc/robot/subsystems/DrivebaseSubsystem.java:
```diff
@@ -9,18 +9,51 @@
+              0,
+              new TrapezoidProfile.Constraints(
+                  MAX_VELOCITY_METERS_PER_SECOND,
+                  0.5 * MAX_VELOCITY_METERS_PER_SECOND))); // FIXME: replace with empirical or
```
Do we want to fix this fixme?

a-viv-a commented on src/main/java/frc/robot/subsystems/DrivebaseSubsystem.java:
```diff
@@ -32,32 +65,60 @@
-  private final AHRS navx = new AHRS(SPI.Port.kMXP, (byte) 200); // NavX connected over MXP
+  /**
+   * The SwerveDriveOdometry class allows us to estimate our vehicle position over time using the
+   * power of :sparkles: math :sparkles:.
```
Do ``:sparkles:`` render in doccomments on vscode?

vecless replied:
Unfortunately, no, I'll remove these 😔

a-viv-a commented on src/main/java/frc/robot/subsystems/DrivebaseSubsystem.java:
```diff
@@ -32,32 +65,60 @@
-  private final SwerveModule frontLeftModule;
-  private final SwerveModule frontRightModule;
-  private final SwerveModule backLeftModule;
-  private final SwerveModule backRightModule;
```
Is there a reason to only store swerve modules in the array? Java object constructors return pointers, so there is no bugs inherent with having both. I think there are times where one is more ergonomic than the other. Thoughts?

vecless replied:
I didn't see a reason to keep too many members in scope. What would be an example of where we'd see benefits from having these named aliases?

a-viv-a commented on src/main/java/frc/robot/subsystems/DrivebaseSubsystem.java:
```diff
@@ -116,11 +187,51 @@ public DrivebaseSubsystem() {
+    this.chassisSpeeds = chassisSpeeds;
+  }
+
+  public void resetGyroAngle(Rotation2d angle) {
```
I have pretty mixed feelings about taking a parameter we do not use... Having an overloaded function in the future, and this function with no parameters now seems like a better path imo.

vecless replied:
Agreed. This was initially written to set the angle adjustment using the àngle param but is no longer doing that. I'll change to update

a-viv-a commented on src/main/java/frc/robot/subsystems/DrivebaseSubsystem.java:
```diff
@@ -134,14 +245,19 @@ public Rotation2d getGyroscopeRotation() {
+    }
+
+    Rotation2d angle;
+    angle = getGyroscopeRotation();
```
nitpick, but can this be one line?

a-viv-a commented on src/main/java/frc/robot/subsystems/DrivebaseSubsystem.java:
```diff
@@ -174,19 +289,22 @@ private void drivePeriodic() {
-  @Override
-  public void periodic() {
+  /**
+   * Based on the current Mode of the drivebase, write the relevant periodic method.
```
would this be run/call, not write? I guess write to the modules? Some methods do more then *just* write.

a-viv-a commented on src/main/java/frc/util/Util.java:
```diff
@@ -1,5 +1,10 @@
+
 package frc.util;
 
+/** Add your docs here. */
```
Should we add them?

a-viv-a commented on src/test/java/com/ironpanthers/lib/UtilTests.java:
```diff
@@ -1,4 +1,4 @@
-package frc.util;
+package com.ironpanthers.lib;
```
Is this intended, or an artifact of moving back to not using packages?

a-viv-a commented on build.gradle:
```diff
@@ -71,6 +72,8 @@ dependencies {
     testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine'
 
+    compileOnly group: "com.google.errorprone", name: "error_prone_annotations", version: "2.3.4"
+    errorprone group: "com.google.errorprone", name: "error_prone_core", version: "2.3.4"
```
are these versions correct? are the guarded annotations and the dependencies on different versioning systems?

a-viv-a commented on src/main/java/com/ironpanthers/lib/Util.java:
```diff
@@ -0,0 +1,44 @@
+package com.ironpanthers.lib;
+
+public class Util {
```
Does this exist twice now? Once here, and once in the non com? Possible artifact of switching back to not using a different library.

vecless replied:
Oops yeah let's drop this one. It becomes an issue if you import both Utils, yeah

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+   * The time that the current trajectory started to be followed. NaN if the trajectory has not been
+   * started yet.
+   */
+  @GuardedBy("trajectoryLock")
```
If there are no threads, do we need guards?

vecless replied:
idt so

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  public final void follow(Trajectory trajectory) {
+    synchronized (trajectoryLock) {
```
Is the synchronized block needed without threads?

vecless replied:
once again, idt so. applies to all future questions of this kind

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+
+  /** Gets the current trajectory that is being followed, if applicable. */
+  public final Optional<Trajectory> getCurrentTrajectory() {
+    synchronized (trajectoryLock) {
```
Is the synchronized block needed without threads?

a-viv-a commented on src/main/java/frc/robot/autonomous/TrajectoryFollower.java:
```diff
@@ -0,0 +1,99 @@
+  public final Optional<T> update(Pose2d currentPose, double time, double dt) {
+    Trajectory trajectory;
+    double timeSinceStart;
+    synchronized (trajectoryLock) {
```
Is the synchronized block needed without threads?

a-viv-a reviewed (changes requested):
There are mostly nitpicks that don't *require* fixes, but the duplicated ``Utils`` and the left over threading stuff probably do / may break things.

vecless committed:
Drop errorprone compiler and references

vecless committed:
Remove unused code

vecless committed:
Clean up DrivebaseSubsystem (docs+unused)

vecless committed:
Remove duplicate Util.java

Also rectifies the location of the UtilTests to match with the `frc.util` pkg it is a part of

vecless committed:
Document TrajectoryFollower a little more

Hopefully this makes it a little clearer how these classes are supposed to work

a-viv-a commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -25,8 +25,6 @@
   @Override
   public void robotInit() {
-    // Instantiate our RobotContainer.  This will perform all our button bindings, and put our
-    // autonomous chooser on the dashboard.
```
Can we leave these comments?

vecless committed:
Merge remote-tracking branch 'origin/dev' into feat/swerve-update-thread

a-viv-a reviewed (approved):
lgtm, lets test on the robot before merge

vecless committed:
Restore comment in robotInit

vecless committed:
Use zeroYaw method of navx to reset gyro

This maintains the logic currently present on `dev`

Brockhold reviewed (approved):
I don't see any tests for the trajectory follower, but I think the logic makes sense. 
Disclaimer: I didn't step through it with examples.

vecless committed:
Document and test trajectory followers

vecless commented:
Should be good (fingers crossed) now. Awaiting test with hardware

Craptop committed:
update constants for module swap

vecless committed:
Merge remote-tracking branch 'origin/fix/practice/module-2-swap' into feat/swerve-update-thread

Craptop committed:
Update odometry logic

vecless committed:
Run :spotlessApply
